Build an image without conda

aperrin66 opened the pull request:
Build a slim image as well as the standard one.

The main difference is that the slim image is not based on Anaconda, which makes it necessary to compile GDAL.

The main gain is to reduce considerably the size of the image (by around 40%).
You can verify this by comparing the sizes of the **latest** and **latest-slim** tags [here](https://hub.docker.com/repository/docker/aperrin66/nansat_base/tags?page=1)

It will also make it easier to work with the image since it won't be necessary to worry about activating the conda environment.


The downsides are that it will be a bit more complex to maintain, and the build time for a first build or when the compilation stage is modified is pretty long: around 20 minutes. Although for a standard build were we can take advantage of the previously built image, it is much quicker (one or two minutes).

@akorosov if you are on board with this, we will probably need to review which compilation options are necessary for GDAL, in particular which file formats need to be supported.

opsdep commented on Dockerfile_slim:
```diff
@@ -0,0 +1,110 @@
+    INSTALL_PREFIX=usr/local \
+    PYTHONUNBUFFERED=1
+
+# Copy artefacts from the builder image
```
I think artifact instead of artefacts in this line is better choice!

akorosov commented:
In general I'm OK with __switching__ to slim image.

But we should make sure that we can run not only Nansat, but also Geospaas with it. I'm especially concerned with conflicts between spatialite in gdal and geo-django. I don't remember exactly, but there were big challenges.

Maybe this __switching__ can be done in two steps: 
1. adding a slim image as a temporary solution for testing of downstream images, 
2. removing conda - based image

We need to add DODS driver:
https://gdal.org/drivers/raster/dods.html

it can be compiled with:
--with-dods-root=${INSTALL_PREFIX}

Some hints here:
https://github.com/conda-forge/gdal-feedstock/blob/master/recipe/build.sh

and libdap needs to be installed. I'm not sure how, here it mentions libdap4:
https://github.com/conda-forge/gdal-feedstock/blob/master/recipe/meta.yaml But that's conda name.

Also need to add `g++` to the slim image.
When I tried to build Nansat using the slim image it said:
```bash
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/local/include -I/usr/local/include/python3.8 -c nansat/pixelfunctions/pixelfunctions.c -o build/temp.linux-x86_64-3.8/nansat/pixelfunctions/pixelfunctions.o -fPIC -Wall -Wno-long-long -pedantic -O3
unable to execute 'gcc': No such file or directory
error: command 'gcc' failed with exit status 1
The command '/bin/sh -c python setup.py install     &&  rm -rf /tmp/{utilities,nansat,setup.py}' returned a non-zero code: 1
```

aperrin66 commented:
Thanks for the comment, I will have a look at all this.

Regarding g++ though, I intentionally left it out of the image, since it's pretty heavy and is only necessary to build the image, but not to run it.
As you found out, it means that it must be re-installed in downstream images.

I had in mind to do something similar as I did here in downstream images: adding a slim version so that we can test the whole chain extensively before deciding to make the switch.

aperrin66 commented:
@akorosov I added a few things:
- the DODS driver
- pinned dependencies versions in the compilation stage
- updated pip packages versions according to what you  did on the standard image
- updated GDAL to 3.1.2 (which required to compile the version 6 of PROJ which is not available as a distribution package in the version of debian used for the python images)

I could build the nansat image on top of this one.

akorosov reviewed (approved):
Great job!
